File: app/src/main/java/dev/thehale/papermc_plugin_template/InventoryDataUtils.java
package dev.thehale.papermc_plugin_template;

import java.util.ArrayList;
import java.util.List;

import org.bukkit.inventory.ItemStack;

final class InventoryDataUtils {

    private InventoryDataUtils() {
        throw new UnsupportedOperationException("Utility class");
    }

    static List<ItemStack> serializeInventoryList(ItemStack[] source) {
        if (source == null) {
            return null;
        }
        List<ItemStack> serialized = new ArrayList<>(source.length);
        for (ItemStack itemStack : source) {
            serialized.add(itemStack == null ? null : itemStack.clone());
        }
        return serialized;
    }

    static ItemStack[] deserializeInventoryList(List<?> source) {
        if (source == null) {
            return null;
        }
        ItemStack[] deserialized = new ItemStack[source.size()];
        for (int i = 0; i < source.size(); i++) {
            Object value = source.get(i);
            deserialized[i] = value instanceof ItemStack itemStack ? itemStack.clone() : null;
        }
        return deserialized;
    }

    static ItemStack[] cloneItemStackArray(ItemStack[] source) {
        if (source == null) {
            return null;
        }
        ItemStack[] clone = new ItemStack[source.length];
        for (int i = 0; i < source.length; i++) {
            clone[i] = source[i] == null ? null : source[i].clone();
        }
        return clone;
    }

    static final class Snapshot {
        private final ItemStack[] contents;
        private final ItemStack[] armor;
        private final ItemStack offhand;

        Snapshot(ItemStack[] contents, ItemStack[] armor, ItemStack offhand) {
            this.contents = contents;
            this.armor = armor;
            this.offhand = offhand;
        }

        ItemStack[] contents() {
            return contents;
        }

        ItemStack[] armor() {
            return armor;
        }

        ItemStack offhand() {
            return offhand;
        }
    }
}

File: app/src/main/kotlin/dev/x208/sheepmerge/InventoryDataUtils.kt
package dev.x208.sheepmerge

import org.bukkit.inventory.ItemStack

internal object InventoryDataUtils {

    @JvmStatic
    fun serializeInventoryList(source: Array<ItemStack?>?): List<ItemStack?>? {
        if (source == null) {
            return null
        }
        return source.map { itemStack -> itemStack?.clone() }
    }

    @JvmStatic
    fun deserializeInventoryList(source: List<*>?): Array<ItemStack?>? {
        if (source == null) {
            return null
        }
        return Array(source.size) { index ->
            val value = source[index]
            if (value is ItemStack) value.clone() else null
        }
    }

    @JvmStatic
    fun cloneItemStackArray(source: Array<ItemStack?>?): Array<ItemStack?>? {
        if (source == null) {
            return null
        }
        return Array(source.size) { index -> source[index]?.clone() }
    }

    class Snapshot(
        private val contents: Array<ItemStack?>?,
        private val armor: Array<ItemStack?>?,
        private val offhand: ItemStack?
    ) {
        fun contents(): Array<ItemStack?>? {
            return contents
        }

        fun armor(): Array<ItemStack?>? {
            return armor
        }

        fun offhand(): ItemStack? {
            return offhand
        }
    }
}
